### DISCUSSION TITLE
Attiny85 TinySoftwareSerial

### DISCUSSION THREAD

**Discussion by g0730n** [SpenceKonde/ATTinyCore#875] [General]
I want to use PB0, and PB1 for Uart, but also be able to switch back within my sketch on attiny85 to control LED connected to those Pins.

I know RX for TinySoftwareSerial uses analog comparator interrupt...

Is there an easy way to use those pins for Serial, then disable that comparator and switch the pins back to general usage, like controlling LED or input?

I saw mention of this around the repository in there but no example on how to do it

Thanks!

**Comment by hmeijdam** [comment]
You don't need tinySoftwareSerial. AttinyCore has built in softwareserial called Serial.

So Serial.begin(xxxx) will start serial communication and if you want to use the pins as GPIO just call Serial.end() and do with the pins what you want. Then call Serial.begin() again to print etc..

**Reply by g0730n** [reply]
Awesome, I will test accordingly!

Reading the documentation on readme, it seemed like it was inferring that TinySoftwareSerial was preferred to use over SoftwareSerial. But then it was also saying using Serial.begin(), or Serial.write(), TinySoftwareSerial is default?

Not sure, but I will try just using "Serial" without including <TinySoftwareSerial.h> and see how it goes, thanks for the input!

**Comment by hmeijdam** [comment]
Something like this.

```
void setup() {
}

void loop() {
Serial.begin(9600);
Serial.println("Hello there");
Serial.end();
pinMode(PB0, OUTPUT);
digitalWrite(PB0, HIGH);
delay(1000);
digitalWrite(PB0, LOW);
delay(1000);
}
```

**Reply by g0730n** [reply]
Yes this is working for me, thank you so much!

**Comment by SpenceKonde** [comment]
All ATTinyCore parts supported by this core will always have a serial port called Serial. If there are hardware serial port(s), Serial will be Serial as God intended it. Otherwise, they will be software serial, an instance of tinySoftSerial. There is no need to include anything for this, it is part of the core to aid in migration of code from stuff with hardware serial.

Just remember that it's being called Serial doesn't change the fact that it's a software serial port, with all the bullshit that normally entails (Fairly limited range of baud rates that work, it's half duplex, all transmit is blocking immediately (rather than running in the background driven by interrupts like normal serial), and  while either sending or receiving, the device cannot do anything else.

**Reply by g0730n** [reply]
I am using default Serial pins for tinycore on 85 running @1mhz, and 2400 baud. Transferring a max of 127 bytes at a time, send to device then have device send back for comfirmation. Have not had any issues yet with these settings.

The two pins I am using for serial are also the 2 LED on device, to save power i am using 10k resistors to power these LED from PB0 and PB1.

Using Serial.begin(2400) and Serial.end() allows me to seamlessly switch between serial transfer and LED control.

Thank you guys for the help!
